service/telemetry.go

```diff
@@ -38,34 +38,27 @@ import (
 	"go.opentelemetry.io/collector/translator/conventions"
 )
 
-// applicationTelemetry is application's own telemetry.
-var applicationTelemetry appTelemetryExporter = &appTelemetry{}
-
-type appTelemetryExporter interface {
-	init(asyncErrorChannel chan<- error, ballastSizeBytes uint64, logger *zap.Logger) error
-	shutdown() error
-}
-
 type appTelemetry struct {
 	views  []*view.View
 	server *http.Server
 }
 
-func (tel *appTelemetry) init(asyncErrorChannel chan<- error, ballastSizeBytes uint64, logger *zap.Logger) error {
+func newAppTelemetry(asyncErrorChannel chan<- error, logger *zap.Logger) (*appTelemetry, error) {
 	level, err := telemetry.GetLevel()
 	if err != nil {
-		return fmt.Errorf("failed to parse metrics level: %w", err)
+		return nil, fmt.Errorf("failed to parse metrics level: %w", err)
 	}
 
-	metricsAddr := telemetry.GetMetricsAddr()
+	tel := &appTelemetry{}
 
+	metricsAddr := telemetry.GetMetricsAddr()
 	if level == configtelemetry.LevelNone || metricsAddr == "" {
-		return nil
+		return tel, nil
 	}
 
-	processMetricsViews, err := telemetry2.NewProcessMetricsViews(ballastSizeBytes)
+	processMetricsViews, err := telemetry2.NewProcessMetricsViews(0)
```

Comment:
What's the consequence of this change? Will we have incorrect metrics?

Reply:
I think we will have correct metrics, because the ballast is also allocated.

Reply:
**If I remember correctly**, the ballast is always allocated in virtual mem. As we never write to it, there are no page faults and it doesn't actually take any memory. From Go's perspective it does consume the mem but from host/container resource perspective, it doesn't. Not accounting for this can produce confusing metrics and force users to account for ballast size on monitoring side. So if a collector actually consumes 2GiB of memory and has a 5GiB of ballast, metrics will show 7GiB consumption while as in reality only 2GiB would be consumed. 

I think this is more of a breaking change than removing the CLI flag as this'll likely cause a lot of alerts in people's environments :)

Reply:
I actually don't think what you say is true (I may be wrong), but I am kind of confident that if it is used in k8s and configured with 4G, and ballast of 5G will OOM. You added that change and I am curios if you tested that.

Reply:
The statement:
> the ballast is always allocated in virtual mem

Any memory allocation is always allocated in virtual memory, independent of accessing or not accessing the memory.

Reply:
I'm sure I tested that (may be incorrectly) but I can test again later today when I start my work day. We never ran into surprises with this or saw unexpected OOMs compared to exported metrics so I think it did work as expected. In any case, I'll confirm again tomorrow and get back.

Reply:
Yea, I meant it **only** stays as virtual and never takes up real mem.

Reply:
Confirmed that you are wrong, and I am wrong, created a small binary and measured the memory using `top` and activity monitor on macos (which both showed very similar results):

```
Ballast Size 8198M - Real Memory Size 267M
Ballast Size 4096M - Real Memory Size 134M
Ballast Size 1024M - Real Memory Size 37M
Ballast Size 128M - Real Memory Size 6M
```

I looked to see if `make` confirms to initialize the memory (which would have been making real memory usage), but could not find that it does it for sure. But I know for sure that `make([]byte, 1000)` will create a slice of a capacity 1000 and 1000 elements which will be 0 initialized (which means it has to allocate some real memory for these). What I think it happens is actually an optimization on the runtime, that it initializes looks like roughly ~4% (estimate) of the elements, so that access to this elements will not trigger page faults and force memory allocation, and the VM will be able to do some kind of prediction (I guess) and allocate the rest of the pages when needed (possibly before being accessed).

Binary is here:

```golang
package main

import (
	"log"
	"runtime"
	"time"
)

const megaBytes = 1024 * 1024

type memoryBallast struct {
	sizeMiB uint32
	ballast []byte
}

func (m *memoryBallast) start() {
	if m.sizeMiB > 0 {
		ballastSizeBytes := uint64(m.sizeMiB) * megaBytes
		m.ballast = make([]byte, ballastSizeBytes)
		log.Printf("Using memory ballast %d", m.sizeMiB)
	}
}

func (m *memoryBallast) shutdown() {
	m.ballast = nil
}

func newMemoryBallast(sizeMiB uint32) *memoryBallast {
	return &memoryBallast{
		sizeMiB: sizeMiB,
	}
}

func main() {
	mb := newMemoryBallast(128)
	mb.start()

	ticker := time.NewTicker(10 * time.Second)
	done := time.After(10 * time.Minute)

	for {
		select {
		case <-done:
			mb.shutdown()
			return
		case <-ticker.C:
			ms := &runtime.MemStats{}
			runtime.ReadMemStats(ms)
			log.Printf("Alloc %d HeapAlloc %d HeadSys %d HeapInUse %d", ms.Alloc, ms.HeapAlloc, ms.HeapAlloc, ms.HeapInuse)
		}
	}
}
```

Reply:
Also proved that HeapAlloc in the memory stats accounts for the entire ballast.

Reply:
That sounds about right. Now that you found it again, I remember the ~3% "overhead". 

> Also proved that HeapAlloc in the memory stats accounts for the entire ballast.
Right. I think that's why we subtracted the size of heap from reported heap size by Go metrics so prom metrics would be closer to reality.

Reply:
What's the conclusion here? If there is only 3% of overhead then supposedly we still need to subtract the remaining 97% which is not using any physical memory? In that case subtracting 100% is more accurate than not subtracting anything?

Reply:
I am not sure I like that we do this in this PR even if it is a bug.

Reply:
Perhaps we can just be more transparent about it and not subtract anything while documenting the effects of using ballast thoroughly so users can decide how they want to consume the metrics. They'll have to account for it in their monitoring/alert systems but probably that is better in the long run. 

My only concern is that this might end of firing a lot of alarms for users if their alerts depend on the affected metrics. 

May be we should continue to subtract 100% during the grace period and warn the users about it with the intension of removing it once we remove the CLI flag.

Reply:
Keep in mind that this is not a guarantee in the golang specs, which means this behavior can change at any moment. So I think us doing this is just wrong, because go 1.16 is free to change this behavior.

We are reallying on an undiffined behavior here which should not be the case in any implementation.

Reply:
@tigrannajaryan this PR can be changed and we can discuss later, there are other 2 PRs that need to be merged first.

Reply:
Also I think we should not be in the business of producing these metrics, we should use OTel library or ask users to use host memory scraper to get this information from the kernel/kublet

Reply:
Agree